notification.go

```diff
@@ -0,0 +1,136 @@
+// SPDX-License-Identifier: ISC
+// Copyright (c) 2019-2021 Bitmark Inc.
+// Use of this source code is governed by an ISC
+// license that can be found in the LICENSE file.
+
+package main
+
+import (
+	"bytes"
+	"encoding/json"
+	"errors"
+	"io/ioutil"
+	"net/http"
+
+	"github.com/gin-gonic/gin"
+	log "github.com/sirupsen/logrus"
+	"github.com/spf13/viper"
+
+	"github.com/bitmark-inc/autonomy-pod-controller/bitcoind"
+)
+
+func (c *Controller) transactionNotify(context *gin.Context) {
+	txID := context.Param("txid")
+	client, err := bitcoind.NewHttpRPCClient(c.httpClient)
+	tx, err := client.GetTransaction(txID)
+	if err != nil {
+		logFields := map[string]interface{}{
+			"txid": txID,
+		}
+		replyWithError(context, err, "failed to get tx from rpc", logFields)
+		return
+	}
+
+	// forloop vins to get addresses
+	vins := make([]string, 0)
+	for _, vin := range tx.Decoded.Vins {
+		addr, err := getVinAddresses(client, vin.TxID, int(vin.Vout))
+		if err != nil {
+			logFields := map[string]interface{}{
+				"txid": vin.TxID,
+			}
+			replyWithError(context, err, "failed to get tx vin address", logFields)
+			return
+		}
+		vins = append(vins, addr...)
+	}
+
+	// forloop vouts to get addresses and value
+	vouts := make([]map[string]interface{}, 0)
+	for _, vout := range tx.Decoded.Vouts {
+		vouts = append(vouts, map[string]interface{}{
+			"addresses": vout.ScriptPubKey.Addresses,
+			"value":     vout.Value,
+		})
+	}
+
+	// prepare the notify params
+	type notifyFormat struct {
+		AccountID string
+		Contents  map[string]string
+		Data      map[string]interface{}
+	}
+	notifyContent := map[string]string{
+		"en": "Transaction Notification",
+	}
+	notifyData := map[string]interface{}{
+		"TxID":          tx.TxID,
+		"Confirmations": tx.Confirmations,
+		"Category":      tx.Details[0].Category,
+		"Amount":        tx.Amount,
+		"Vins":          vins,
+		"Vouts":         vouts,
+	}
+
+	body := new(bytes.Buffer)
+	if err := json.NewEncoder(body).Encode(notifyFormat{AccountID: c.ownerDID, Contents: notifyContent, Data: notifyData}); err != nil {
+		logFields := map[string]interface{}{
+			"notifyData": notifyData,
+		}
+		replyWithError(context, err, "failed to encode tx", logFields)
+		return
+	}
+
+	// start to call notification api
+	notifyURL := viper.GetString("messaging.endpoint") + "/api/accounts/notification"
```

Comment:
`messaging.endpoint` should be used only for the messaging module. I might configure this as a `notifictaion.tx_callback_url`.